provisioning/device/src/tpm_registration.ts

```diff
@@ -0,0 +1,239 @@
+              } else {
+                this._fsm.handle('activateSessionKey', tpmChallenge, registrationInfo, registerCallback);
+              }
+            });
+          },
+          activateSessionKey: (tpmChallenge, registrationInfo, registerCallback) => {
+            /*Codes_SRS_NODE_DPS_TPM_REGISTRATION_16_004: [The `register` method shall store the session key in the TPM by calling the `activateIdentityKey` method of the `TpmSecurityClient` object passed to the constructor with the following arguments:
+            - `sessionKey`: the session key property of the `TpmChallenge` object returned by the previous call to `TpmProvisioningTransport.getAuthenticationChallenge`
+            - a callback that will handle an optional error if the operation fails.]*/
+            this._securityClient.activateIdentityKey(tpmChallenge.authenticationKey, (err) => {
+              if (err) {
+                debug('failed to activate the sessionKey: ' + err.toString());
+                /*Codes_SRS_NODE_DPS_TPM_REGISTRATION_16_010: [If any of the calls the the `TpmSecurityClient` or the `TpmProvisioningTransport` fails, the `register` method shall call its callback with the error resulting from the failure.]*/
+                this._fsm.transition('notStarted', err, registerCallback);
+              } else {
+                this._fsm.handle('createRegistrationSas', registrationInfo, registerCallback);
+              }
+            });
+          },
+          createRegistrationSas: (registrationInfo, registerCallback) => {
+            this._createRegistrationSas(registrationInfo.registrationId, (err, sasToken) => {
+              if (err) {
+                debug('failed to get sign the initial authentication payload with the sessionKey: ' + err.toString());
+                /*Codes_SRS_NODE_DPS_TPM_REGISTRATION_16_010: [If any of the calls the the `TpmSecurityClient` or the `TpmProvisioningTransport` fails, the `register` method shall call its callback with the error resulting from the failure.]*/
+                  this._fsm.transition('notStarted', err, registerCallback);
+                } else {
+                this._fsm.transition('registrationInProgress', registrationInfo, sasToken, registerCallback);
+              }
+            });
+          },
+          cancel: (callback) => {
+            /*Codes_SRS_NODE_DPS_TPM_REGISTRATION_16_011: [The `cancel` method shall interrupt the ongoing registration process.]*/
+            this._transport.cancel((err) => {
+              if (err) {
+                debug('failed to stop provisioning transport: ' + err.toString());
+              }
+              this._securityClient.cancel((err) => {
+                if (err) {
+                  debug('failed to stop provisioning transport: ' + err.toString());
+                }
+                this._fsm.transition('notStarted', err, callback);
+              });
+            });
+          },
+          '*': () => this._fsm.deferUntilTransition()
```

Comment:
>'*': () => this._fsm.deferUntilTransition() [](start = 10, length = 43)

This is used in some states but not others.  Is this by design?  This relates to my question about expected behavior if the user calls register if a registration is in progress.

Reply:
it should be by design. If there is no "catch-all" clause like this one, the state machine doesn't error out, it just doesn't do anything.

---
In reply to: [152068942](https://github.com/Azure/azure-iot-sdk-node/pull/155#discussion_r152068942) [](ancestors = 152068942)